File: frontend/src/stores/incidents.ts
import { defineStore } from 'pinia'
import { api } from '../api/http'
import type { Incident, WsMessage } from '../types'

export const useIncidentsStore = defineStore('incidents', {
  state: () => ({
    byService: {} as Record<string, Incident[]>,
    loading: {} as Record<string, boolean>,
  }),

  actions: {
    async fetchForService(id: string) {
      if (this.loading[id]) return
      this.loading[id] = true
      try {
        this.byService[id] = await api.fetchIncidents(id)
      } finally {
        this.loading[id] = false
      }
    },

    applyOpened(msg: Extract<WsMessage, { type: 'incident_opened' }>) {
      const list = this.byService[msg.service_id]
      if (!list) return
      const incident: Incident = {
        id: msg.incident_id,
        service_id: msg.service_id,
        started_at: msg.started_at,
        resolved_at: null,
        status: 'open',
        trigger_status: msg.trigger_status,
        notes: null,
      }
      this.byService[msg.service_id] = [incident, ...list]
    },

    applyResolved(msg: Extract<WsMessage, { type: 'incident_resolved' }>) {
      const list = this.byService[msg.service_id]
      if (!list) return
      const incident = list.find(i => i.id === msg.incident_id)
      if (incident) {
        incident.resolved_at = msg.resolved_at
        incident.status = 'resolved'
      }
    },

    async resolve(serviceId: string, incidentId: string, notes?: string) {
      const updated = await api.resolveIncident(serviceId, incidentId, notes)
      const list = this.byService[serviceId]
      if (list) {
        const idx = list.findIndex(i => i.id === incidentId)
        if (idx !== -1) list[idx] = updated
      }
    },
  },
})

File: backend/src/routes/incidents.rs
use axum::{
    extract::{Path, State},
    Json,
};
use chrono::Utc;
use crate::{
    error::AppError,
    models::incident::{Incident, ResolveIncident},
    state::AppState,
    ws::messages::WsMessage,
};

pub async fn list_incidents(
    State(state): State<AppState>,
    Path(service_id): Path<String>,
) -> Result<Json<Vec<Incident>>, AppError> {
    let incidents = sqlx::query_as::<_, Incident>(
        "SELECT id, service_id, started_at, resolved_at, status, trigger_status, notes
         FROM incidents WHERE service_id = $1
         ORDER BY started_at DESC LIMIT 100",
    )
    .bind(&service_id)
    .fetch_all(&state.db)
    .await?;

    Ok(Json(incidents))
}

pub async fn resolve_incident(
    State(state): State<AppState>,
    Path((service_id, incident_id)): Path<(String, String)>,
    Json(body): Json<ResolveIncident>,
) -> Result<Json<Incident>, AppError> {
    let now = Utc::now();

    let rows = sqlx::query(
        "UPDATE incidents SET resolved_at = $1, status = 'resolved', notes = $2
         WHERE id = $3 AND service_id = $4 AND status = 'open'",
    )
    .bind(now)
    .bind(&body.notes)
    .bind(&incident_id)
    .bind(&service_id)
    .execute(&state.db)
    .await?
    .rows_affected();

    if rows == 0 {
        return Err(AppError::NotFound);
    }

    let incident = sqlx::query_as::<_, Incident>(
        "SELECT id, service_id, started_at, resolved_at, status, trigger_status, notes
         FROM incidents WHERE id = $1",
    )
    .bind(&incident_id)
    .fetch_one(&state.db)
    .await?;

    let _ = state.tx.send(WsMessage::IncidentResolved {
        incident_id: incident.id.clone(),
        service_id: service_id.clone(),
        resolved_at: now,
    });

    Ok(Json(incident))
}
